### PR TITLE
StylesPreview: Fix endless loop of ratio calculations when on the threshold of a scrollbar

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Similar issue to #55112

Fix the `StylesPreview` component so that it doesn't endlessly "dance" when the viewport height means that the preview's container is right at the threshold of a scrollbar appearing.

## Why?

As described in https://github.com/WordPress/gutenberg/issues/55112#issuecomment-1752390866 the ratio calculations in the `StylesPreview` component can easily wind up in an endless loop when a scrollbar is shown or hidden. It goes something like this:

1. The sidebar is rendered without any previews
2. Enough previews are rendered to cause a scrollbar to appear
3. The presence of the scrollbar causes the sizes of the previews to be recalculated
4. The previews are re-rendered, and now that they're smaller, result in no scrollbar being present
5. As no scrollbar is present, the container is now wider, so the previews can be rendered larger again
6. The previews are rendered larger, which introduces the scrollbar (we are now back at step 2, so the rest is an infinite loop)

This issue will only be apparent at very particular viewport heights, and depending on the number of previews present. I could reproduce this issue reliably in Chrome on Mac at the following viewport heights:

* Styles variations picker in global styles, with TT4 theme active: `568px` viewport height
* Main global styles panel in the righthand sidebar: `663px` viewport height

## How?

TL;DR — don't recalculate ratio for these previews if the change in width is only quite small (and not worth the penalty of an infinite loop)

Long story:

The fix is a little convoluted, but required a few steps to avoid an infinite loop, and keep things performant (as far as I can tell):

* Use an initial ratio calculation as on trunk to ensure that the first renders of the previews are at the correct size.
* Then, use a throttled update to set state to store the calculated ratio. In this throttled logic, set it to update the state at both the leading and trailing edges of the throttle so that the state is updated as soon as possible since the first render, and is then throttled (at `250ms`).
* Once the ratio state has been updated by the throttled function, use this state for the ratio.
* Only update the ratio state again if the ratio has change by greater than `0.1` — this is a big enough threshold to account for the show/hiding of the scrollbar, without having a negative effect (subjectively) on the appearance of the previews

While I was at it, I've also fixed an issue with the preview iframes not filling their containers in mobile viewports.

## Testing Instructions

1. On `trunk` see if you can replicate the dancing previews issue by using the viewport heights from the "Why?" section above. The key screens to look at are the root of global styles in the right hand sidebar, and the style variations screen.
2. With this PR applied, there shouldn't be any dancing style previews at any viewport sizes.
3. Double-check that the previews still look good at mobile screen sizes.
4. Also, double-check that the Styles screen in the global styles browse mode (the dark background screen) work as expected

### Global styles (before — trunk)

https://github.com/WordPress/gutenberg/assets/14988353/ca817c7a-f0b8-42a2-9037-ba955b53601f

### Style variations (before — trunk)

https://github.com/WordPress/gutenberg/assets/14988353/c7df1436-c0c8-479e-8634-770c1b605a59

### Global styles (with this PR applied)

https://github.com/WordPress/gutenberg/assets/14988353/ce1c8d49-e833-40e1-a724-f0f2421b7998

### Style variations (with this PR applied)

https://github.com/WordPress/gutenberg/assets/14988353/8ddd0378-9a14-4753-9433-c245b6a56e3e

| Mobile viewport sizing (before) | Mobile viewport sizing (after) |
| --- | --- |
| image |  image |

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/preview.js [hunk 1]
```diff
@@ -78,8 +89,49 @@ const StylesPreview = ( { label, isFocused, withHoverView } ) => {
 	const [ styles ] = useGlobalStylesOutput();
 	const disableMotion = useReducedMotion();
 	const [ isHovered, setIsHovered ] = useState( false );
-	const [ containerResizeListener, { width } ] = useResizeObserver();
-	const ratio = width ? width / normalizedWidth : 1;
+	const [ containerResizeListener, sizes ] = useResizeObserver();
+	const { width } = sizes || { width: 0, height: 0 };
```
Comment (ramonjd) [inline packages/edit-site/src/components/global-styles/preview.js]: Not sure if it matters here since there's no `useMemo` or similar relying on the constant ref, but same as for `THROTTLE_OPTIONS`, could `{ width: 0, height: 0 }` be defined outside of the component as well since other defaults are?

**Review by ramonjd — APPROVED** [review]
Tested in Firefox, Chrome, Safari and Edge

✅  Can confirm it fixes the dancing styles preview panes. I could only reproduce on Chrome
✅  Also, the borders of the style variation previews look good on all browser in mobile view

LGTM

Thanks for going deep on this one, and removing a weird bug. 

Do you think much of the workaround effort could be spared if we removed 'motion.div'? I'm not sure, but I'm assuming it's handling animations and therefore layout.

Just wondering if the old fashioned `iframe.onLoad` + `frame.document.body.offsetHeight` would be simpler 🤷🏻

**Comment by ramonjd** [comment]
htph

File: packages/edit-site/src/components/global-styles/preview.js (hunk 1 above)
Comment (andrewserong) [reply packages/edit-site/src/components/global-styles/preview.js]: Whoops! I forgot to revert this line, this one should've been `const [ containerResizeListener, { width } ] = useResizeObserver();`. Good catch! Reverted that line in https://github.com/WordPress/gutenberg/pull/57090/commits/32a33feba5326338c70de22e32f434f96638c017

**Comment by andrewserong** [comment]
Thanks for the review! 🙇 

> Do you think much of the workaround effort could be spared if we removed 'motion.div'? I'm not sure, but I'm assuming it's handling animations and therefore layout.
> Just wondering if the old fashioned iframe.onLoad + frame.document.body.offsetHeight would be simpler 🤷🏻

Good questions! It seems some of the animations are a bit dependent on `motion.div`, and at first I didn't have much success at removing them without causing other visual issues. Definitely setting the offsetHeight onLoad would be simpler, but there's then the potential issue of a user resizing the window in narrow viewports, or what happens if or when the styles are in a non-fixed-width container. There's also a few calculations happening throughout the inside of the component, so it's a little challenging to untangle. But I could be overthinking it!

I do think this component is probably doing a little too much heavy lifting when it comes to determining ratios. Like perhaps another way of doing it would be to have a couple of hard-coded ratios that the component uses depending on a threshold size, so instead of a whole bunch of multiplication, we could just swap out one or two classes (or constants) or something?

Still, it's an interesting problem when trying to get a component to perform changes reactively to the container width. I suspect there's likely one or two other components that might have this problem, too.

For now, I'll leave the implementation like this (and since it's the end of the week). If there's no objections, I'll merge this one in on Monday, and happy to look into follow-ups to go for a simpler / stripped down approach!
